tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -0,0 +1,242 @@
+//          and the allocation size of M is [1, M/num_devices_,N]
+// Up to some broadcast and view ops, a straightforward program to generate the
+// output could be summarized as
+//     | C_unreduced = pointwise_multiply(A,B)
+//     | C_locally_reduce = reduction(C_unreduced, axis=`K/num_devices_`, op=sum)
+//     | C = reduce_scatter(C_unreduced, op=sum)
+// where:
+// - C has unsharded size [M,num_devices_,K/num_devices_,N],
+//    and is sharded on `num_devices_`
+// - C_locally_reduce has unsharded size [M,num_devices_,N],
+//    and is sharded on `num_devices_`
+// - C has unsharded size [num_devices_, M/num_devices_, N]
+//    and is sharded on `num_devices_`
+//
+// We want to compare this baseline program with one that is functionnally
+// identical but achieves more overlap between computations and communications.
+// Our goal is to interlave the comms and compute using a technic called
+// "reduce-scatter based pipelining". To do so, we further split the row
+// dimension M with a factor `S` representing the number of tiles, and we apply
+// the operations successively on tensors slices accross S, changing stream at
+// each iteration. Assuming the following shapes:
+//     - A [S, num_devices_, M/S, K/num_devices_], sharded on num_devices_
+//     - B [num_devices_, K/num_devices_, N], sharded on num_devices_
+//     - C [S, num_devices_, M/(S*num_devices_), N], sharded on num_devices_
+// the program could be summarized as:
+//     | for (j=0; j<S; j++) {
+//     |   setCurrentStream(Stream[j])
+//     |   C_unreduced[j] = pointwise_multiply(A[j],B)
+//     |   C_locally_reduce[j] = local_reduction(C_unreduced[j], axis=`K/num_devices_`, op=sum)
+//     |   C[j]=reduce_scatter(C_locally_reduce[j], op=sum)
+//     | }
+// where "[j]" referes to taking a slice onto the `S` dimension.
+// Remarks:
+//   1) it is convenient to have "S" as being the outermost dimension so
+//      C_locally_reduce[j] is a contiguous buffer.
+//   2) The layout needs to match
+//      the reduce-scatter semantics, i.e., the first dimension is reduced and
+//      the second is scattered. This is why we choose the layouts to be
+//      [S, sharded_axis, M, ...]
+// clang-format on
+TEST_F(OverlapTest, SimpleComputeComm) {
+  std::vector<c10::cuda::CUDAStream> streams;
+  for (auto j : c10::irange(params.S)) {
+    // define the sliced tensors
+    auto ta_j = getSlice(ta_, j);
+    auto tc_locally_reduced_j = getSlice(tc_locally_reduced_, j);
+    auto tc_j = getSlice(tc_, j);
+
+    if (params.use_different_streams) {
+      auto new_stream = c10::cuda::getStreamFromPool(
+          /*isHighPriority=*/true, my_device_index_);
+      streams.push_back(new_stream);
+      setCurrentCUDAStream(new_stream);
+    }
+
+    // local compute
+    computeATen(ta_j, tb_, tc_locally_reduced_j);
+    // communication
+    world_communicator_->_reduce_scatter_base(tc_j, tc_locally_reduced_j)
+        ->wait();
```

Comment:
Is this `wait` necessary?

Reply:
Yes, it is always necessary to `wait()` after a communication is posted. However, note here that we change stream at each iteration, and so `wait()` is the last kernel posted on the stream. Therfore, the synchronization will occur only after all the kernels and comms are executed, when we synchronize the stream with the main stream, just before the validation.